overlord/devicestate/devicemgr.go

```diff
@@ -0,0 +1,384 @@
+	}
+
+	var retrieveSerialID string
+	err = t.Get("serial-setup-task", &retrieveSerialID)
+	if err != nil {
+		return err
+	}
+	retrieveSerial := st.Task(retrieveSerialID)
+	if retrieveSerial == nil {
+		return fmt.Errorf("internal error: cannot find retrieve serial task")
+	}
+
+	pk, err := keyPair(st)
+	if err != nil {
+		return err
+	}
+
+	cli := &http.Client{Timeout: 30 * time.Second}
+
+	st.Unlock()
+	resp, err := cli.Get(serialRequestURL)
+	st.Lock()
+	if err != nil {
+		return &state.Retry{After: 60 * time.Second}
+	}
+	if resp.StatusCode >= 400 && resp.StatusCode < 500 {
+		return fmt.Errorf("unexpected status code %d trying to get serial request nonce/ticket (from %s)", resp.StatusCode, serialRequestURL)
+	}
+	if resp.StatusCode != 200 {
+		return &state.Retry{After: 60 * time.Second}
+	}
+
+	dec := json.NewDecoder(resp.Body)
+	var nonceTicket nonceTicketResp
+	err = dec.Decode(&nonceTicket)
+	if err != nil { // assume broken i/o
+		return &state.Retry{After: 60 * time.Second}
+	}
+
+	// XXX: do it the hard way for now (should be much simpler soon)
+	privKey := asserts.OpenPGPPrivateKey(packet.NewRSAPrivateKey(time.Unix(1, 0), pk))
+	encodedPubKey, err := asserts.EncodePublicKey(privKey.PublicKey())
+	if err != nil {
+		return fmt.Errorf("internal error: cannot encode device public key: %v", err)
+
+	}
+
+	serialReq, err := asserts.FreestandingSign(asserts.SerialRequestType, map[string]interface{}{
+		"brand-id":     devIdentity.Brand,
+		"model":        devIdentity.Model,
+		"nonce-ticket": nonceTicket.NonceTicket,
+		"device-key":   string(encodedPubKey),
+	}, nil, privKey) // XXX: fill body with some agreed hardware details
+	if err != nil {
+		return err
+	}
+
+	encodedSerialReq := asserts.Encode(serialReq)
+
+	// NB: until we get at least an Accepted (202) we need to
+	// retry from scratch creating a new nonce because the
+	// previous one used could have expired
+
+	st.Unlock()
+	serial, err := submitSerialRequest(cli, encodedSerialReq)
+	st.Lock()
+	if err == errPoll {
+		retrieveSerial.Set("serial-setup", serialSetup{
+			SerialRequest: string(encodedSerialReq),
+		})
+		// TODO: delay next task?
+		t.SetStatus(state.DoneStatus)
+		return nil
+	}
+	if err != nil { // errors and retries
+		return err
+	}
+
+	retrieveSerial.Set("serial-setup", serialSetup{
+		Serial: string(asserts.Encode(serial)),
+	})
+	t.SetStatus(state.DoneStatus)
+	return nil
+}
+
+func doRetrieveAndInstallSerial(t *state.Task, _ *tomb.Tomb) error {
+	st := t.State()
+	st.Lock()
+	defer st.Unlock()
+
+	devIdentity, err := auth.Device(st)
+	if err != nil {
+		return err
+	}
+
+	var serialSup serialSetup
+	err = t.Get("serial-setup", &serialSup)
+	if err != nil {
+		return err
+	}
+
+	var serial *asserts.Serial
+
+	if serialSup.SerialRequest != "" {
+		// delivery of serial-request was asked to poll
+		cli := &http.Client{Timeout: 30 * time.Second}
+		var err error
+		st.Unlock()
+		serial, err = submitSerialRequest(cli, []byte(serialSup.SerialRequest))
```

Comment:
Again feels like unnecessary boilerplate. I wonder how a task that does the whole thing start-to-finish, skipping unnecessary steps done before based on task state, would feel like. Note this is not even a "Setup" of sorts, as we don't need to have this as external input. This is simply state inside the task itself, similar to what we do elsewhere for tracking old state, etc.